### PR TITLE
Added PLYWriter::writeBinary() ostream overload 

### PR DESCRIPTION
**Pull request by kaipomauli**
modified existing PLYWriter::writeBinary() to open file and then call new overload for data aggregation

added PLYWriter::writeBinary() overload that adds data to referenced ofstream buffer

### CODE REVIEW

File: io/include/pcl/io/ply_io.h [hunk 1]
```diff
@@ -591,6 +591,18 @@ namespace pcl
...
+	  /** \brief Save point cloud data to a PLY file containing n-D points, in BINARY format
+	   * \param[in] fs the output buffer
+	   * \param[in] cloud the point cloud data message
+	   * \param[in] origin the sensor data acquisition origin
+	   * (translation) \param[in] orientation the sensor data acquisition origin
+	   * (rotation) \param[in] use_camera if set to true then PLY file will use element
+	   * camera else element range_grid will be used
+	   */
+	  int
+	  writeBinary(std::ofstream& fs,const pcl::PCLPointCloud2& cloud,
```
[the remark this answers is not in the document]
Comment (kaipomauli) [reply io/include/pcl/io/ply_io.h]: Hi @mvieth after the build failed I only noticed the change from std::ofstream to std::ostream. So we need the overload to output to a general stream and not specifically a file stream, yes?

**Review by mvieth — DISMISSED** [review]
This results in a small behaviour change for `writeBinary` (saving to file), namely that on Windows, the ply header will have Unix-style line endings (only line-feed), whereas it previously had Windows-style line endings (carriage-return + line-feed). This is because previously, the file was opened in text mode (where cr+lf are used on Windows), then the ply header was written, then reopened in binary mode. Now, it is only opened in binary mode, so no carriage-return is inserted.
I think this change makes sense. I checked all ply files from different sources I had on my computer, and only a single one uses Windows-style line endings in the ply header (an old ply file from the opencv_contrib repository). All others use Unix-style line endings, so I think PCL should do the same.

To sum up, the changes look good to me :smile: Thank you, I will merge this pull request in a few days.

File: io/include/pcl/io/ply_io.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mvieth) [reply io/include/pcl/io/ply_io.h]: Yes, `std::ostream` is more useful because then e.g. also `std::stringstream` can be used

**Comment by kaipomauli** [comment]
Hey, I assumed one of the tests that read ply files would catch a discrepancy in formating so I did not dig that much further XD Awesome I'll get onto another issue.

File: io/src/ply_io.cpp [hunk 2]
```diff
@@ -1209,14 +1209,36 @@ pcl::PLYWriter::writeBinary (const std::string &file_name,
...
+}
+
+int
+pcl::PLYWriter::writeBinary (std::ostream& fs,
+                             const pcl::PCLPointCloud2& cloud,
+                             const Eigen::Vector4f& origin,
+                             const Eigen::Quaternionf& orientation,
+                             bool use_camera)
+{
+  fs.imbue(std::locale::classic());
```
Comment (larshg) [inline io/src/ply_io.cpp:1239]: I assume this makes use of a default locale independent on the user settings? Do we do the same when we read? Does it even matter in binary mode?

File: io/src/ply_io.cpp [hunk 3]
```diff
@@ -1209,14 +1209,36 @@ pcl::PLYWriter::writeBinary (const std::string &file_name,
...
+  {
+    fs.close();
+    return -1;
+  }
+  fs.close();
+  return 0;
+}
+
+int
+pcl::PLYWriter::writeBinary (std::ostream& fs,
```
Comment (larshg) [inline io/src/ply_io.cpp:1233]: nitpick - since its not necessarily a file stream, maybe another abbreviation would be appropriate? os?

Comment (kaipomauli) [reply io/src/ply_io.cpp:1233]: Absolutely. I will make the suggested change

File: io/src/ply_io.cpp (hunk 2 above)
Comment (kaipomauli) [reply io/src/ply_io.cpp:1239]: Unsure about this, but I found we are doing the same thing in pcl::PCDWriter::writeBinary as well as pcl::PLYWriter::generateHeader (albeit we are writing ASCII characters to an ostringstream there). Setting the locale does seem to matter even in binary mode (see https://stackoverflow.com/questions/1831369/writing-binary-files-using-c-does-the-default-locale-matter). I also imagine TEST (PCL, PLYReaderWriter) would catch an issue with the number format
